### ISSUE TITLE
Getting UnsatisfiedLinkError: undefined symbol: _ZN2td16verbosity_binlogE while running java example (CentOS 7)

### ISSUE THREAD

**Issue by UristMcKerman** [tdlib/td#850]
TL; DR: build goes without visible errors, but in the end contents of libtddb.a are not linked to libtdjni.so. What I can do to fix the situation?

Did everything following the guide https://github.com/tdlib/td/blob/master/example/java/README.md

```
[jnibuild]# cmake -DCMAKE_BUILD_TYPE=Release -DTD_ENABLE_JNI=ON -DCMAKE_INSTALL_PREFIX:PATH=../example/java/td ..
-- Could NOT find ccache (this is NOT an error)
-- Found OpenSSL: /usr/local/ssl/include /usr/local/ssl/lib/libssl.so;/usr/local/ssl/lib/libcrypto.so
-- Found ZLIB: /usr/include /usr/lib64/libz.so
-- Found JNI: /opt/jdk1.8.0_101/include;/opt/jdk1.8.0_101/include/linux;/opt/jdk1.8.0_101/include /opt/jdk1.8.0_101/jre/lib/amd64/libjawt.so;/opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so
-- Could NOT find Readline (missing:  READLINE_INCLUDE_DIR READLINE_LIBRARY)
-- Could NOT find Readline (this is NOT an error)
-- Configuring done
-- Generating done
-- Build files have been written to: build

[jnibuild]# cmake --build . --target install
[jnibuild]# cd ../example/java/build

[build]# cmake -DCMAKE_BUILD_TYPE=Release -DTd_DIR=../td/lib/cmake/Td -DCMAKE_INSTALL_PREFIX:PATH=.. ..
-- Found JNI: /opt/jdk1.8.0_101/include;/opt/jdk1.8.0_101/include/linux;/opt/jdk1.8.0_101/include /opt/jdk1.8.0_101/jre/lib/amd64/libjawt.so;/opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so
-- Found Java: /opt/jdk1.8.0_101/bin/javac /opt/jdk1.8.0_101/bin/javadoc
-- Configuring done
-- Generating done
-- Build files have been written to: example/java/build

[build]# cmake --build . --target install
[ 20%] Generating Java TDLib API source files
[ 20%] Built target td_generate_java_api
[ 40%] Building Java code
[ 40%] Built target build_java
[ 60%] Generating Javadoc documentation
...
Scanning dependencies of target tdjni
[ 80%] Building CXX object CMakeFiles/tdjni.dir/td_jni.cpp.o
[100%] Linking CXX shared library libtdjni.so
[100%] Built target tdjni
Install the project...
-- Install configuration: "Release"
-- Installing: /root/td-master/example/java/build/../bin/libtdjni.so
-- Removed runtime path from "../bin/libtdjni.so"

[build]# ls ../bin
libtdjni.so

```
Then I see only libtdjni.so, no classes, so I assume build_java did not work properly, but that's not a problem, I compile classes myself:
`[java]# javac -d ./bin org/drinkless/tdlib/example/Example.java org/drinkless/tdlib/Client.java org/drinkless/tdlib/Log.java org/drinkless/tdlib/TdApi.java`

And then when I launch application I see an error
```
[bin]# java '-Djava.library.path=.' org/drinkless/tdlib/example/Example
java.lang.UnsatisfiedLinkError: example/java/bin/libtdjni.so: /root/td-master/example/java/bin/libtdjni.so: undefined symbol: _ZN2td16verbosity_binlogE
        at java.lang.ClassLoader$NativeLibrary.load(Native Method)
        at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1857)
        at java.lang.Runtime.loadLibrary0(Runtime.java:870)
        at java.lang.System.loadLibrary(System.java:1122)
        at org.drinkless.tdlib.example.Example.<clinit>(Example.java:58)
Exception in thread "main" java.lang.UnsatisfiedLinkError: org.drinkless.tdlib.Client.nativeClientExecute(Lorg/drinkless/tdlib/TdApi$Function;)Lorg/drinkless/tdlib/TdApi$Object;
        at org.drinkless.tdlib.Client.nativeClientExecute(Native Method)
        at org.drinkless.tdlib.Client.execute(Client.java:103)
        at org.drinkless.tdlib.example.Example.main(Example.java:297)

```
This symbol is defined in example/java/td/lib/libtddb.a:
```
[lib]# nm libtddb.a | grep binlogE
                 U _ZN2td16verbosity_binlogE
```

So, as far as I can see, there is a problem with linking contents of libtddb.a to libtdjni.so.

**Comment by levlam** [comment]
Everything is linked automatically and there is no way on successful build to not link something. Could you paste output of `cmake --build . --target install` and full output of cmake configuration invocations somewhere?

**Comment by UristMcKerman** [comment]
Here the part that matters
```
[ 60%] Building Java code
/opt/jdk1.8.0_101/bin/javac -d /root/td-master/example/bin /root/td-master/example/java/org/drinkless/tdlib/example/Example.java /root/td-master/example/java/org/drinkless/tdlib/Client.java /root/td-master/example/java/org/drinkless/tdlib/Log.java /root/td-master/example/java/org/drinkless/tdlib/TdApi.java
gmake[2]: Leaving directory `/root/td-master/example/java/build'
[ 60%] Built target build_java
/usr/bin/gmake -f CMakeFiles/tdjni.dir/build.make CMakeFiles/tdjni.dir/depend
gmake[2]: Entering directory `/root/td-master/example/java/build'
cd /root/td-master/example/java/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /root/td-master/example/java /root/td-master/example/java /root/td-master/example/java/build /root/td-master/example/java/build /root/td-master/example/java/build/CMakeFiles/tdjni.dir/DependInfo.cmake --color=
Scanning dependencies of target tdjni
gmake[2]: Leaving directory `/root/td-master/example/java/build'
/usr/bin/gmake -f CMakeFiles/tdjni.dir/build.make CMakeFiles/tdjni.dir/build
gmake[2]: Entering directory `/root/td-master/example/java/build'
[ 80%] Building CXX object CMakeFiles/tdjni.dir/td_jni.cpp.o
/opt/rh/devtoolset-3/root/usr/bin/g++   -DPACKAGE_NAME=\"org/drinkless/tdlib\" -Dtdjni_EXPORTS -std=c++1y  -O3 -DNDEBUG -fPIC -isystem /opt/jdk1.8.0_101/include -isystem /opt/jdk1.8.0_101/include/linux -isystem /usr/local/include    -std=c++14 -o CMakeFiles/tdjni.dir/td_jni.cpp.o -c /root/td-master/example/java/td_jni.cpp
[100%] Linking CXX shared library libtdjni.so
/usr/bin/cmake -E cmake_link_script CMakeFiles/tdjni.dir/link.txt --verbose=1
/opt/rh/devtoolset-3/root/usr/bin/g++  -fPIC -std=c++1y  -O3 -DNDEBUG  -shared -Wl,-soname,libtdjni.so -o libtdjni.so CMakeFiles/tdjni.dir/td_jni.cpp.o /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so /usr/local/lib/libtdclient.a -ltdcore /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so -Wl,-rpath,/opt/jdk1.8.0_101/jre/lib/amd64/server:
gmake[2]: Leaving directory `/root/td-master/example/java/build'
[100%] Built target tdjni
gmake[1]: Leaving directory `/root/td-master/example/java/build'
/usr/bin/cmake -E cmake_progress_start /root/td-master/example/java/build/CMakeFiles 0
/usr/bin/gmake -f CMakeFiles/Makefile2 preinstall
gmake[1]: Entering directory `/root/td-master/example/java/build'
gmake[1]: Nothing to be done for `preinstall'.
gmake[1]: Leaving directory `/root/td-master/example/java/build'
Install the project...
/usr/bin/cmake -P cmake_install.cmake
-- Install configuration: "Release"
-- Installing: /root/td-master/example/java/build/../bin/libtdjni.so
-- Removed runtime path from "../bin/libtdjni.so"
```

And apparently yes, javac outputs to /example/bin instead of /example/java/bin, which is not a big deal.

**Comment by UristMcKerman** [comment]
[build/CMakeFiles/CMakeOutput.log is here](https://pastebin.com/AeffUXWh)

**Comment by levlam** [comment]
Could you try to follow instructions from https://tdlib.github.io/td/build.html?language=Java from scratch from clonning again TDLib?

**Comment by UristMcKerman** [comment]
Yes, I followed instructions from Java->Linux->Other, no debug, install in ./td/tdlib, g++  - with exactly same result.

**Comment by UristMcKerman** [comment]
AFAIK the issue is in example/java/build/CMakeFiles/tdjni.dir/link.txt not linking other .a files as it should. I will try to link them manually

**Comment by levlam** [comment]
@UristMcKerman 
```
[lib]# nm libtddb.a | grep binlogE
                 U _ZN2td16verbosity_binlogE
```
nm tells that the symbol is undefined even in libtddb.a, despite it is defined in tddb/td/db/BinlogEvent.cpp and must be there.

**Comment by UristMcKerman** [comment]
It's is just my mistake, full output was:

[root@dev01 jnibuild]# nm tddb/libtddb.a | grep verbosity_binlogE
                 U _ZN2td16verbosity_binlogE
0000000000000000 D _ZN2td16verbosity_binlogE

But when looking at libtdjni.so I get

[root@dev01 td]# nm tdlib/bin/libtdjni.so | grep verbosity_binlogE
                 U _ZN2td16verbosity_binlogE

**Comment by UristMcKerman** [comment]
Hello! I've retruned to work on this issue and I'm completely lost: I've tracked down this symbol from the begining to the end and figured out that it is not included while executing command

`g++ -fPIC -O3 -DNDEBUG  -shared -Wl,-soname,libtdjni.so -o libtdjni.so CMakeFiles/tdjni.dir/td_jni.cpp.o /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so /usr/local/lib/libtdclient.a -ltdcore /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so -Wl,-rpath,/opt/jdk1.8.0_101/jre/lib/amd64/server:`

Firstly, the content of td/example/java/td/lib/* where all .a files lie are not included. But, even when I explicitly tell g++ to link these libs

`/opt/rh/devtoolset-3/root/usr/bin/g++ -D_FILE_OFFSET_BITS=64 -pthread -std=c++14 -fPIC -O3 -DNDEBUG  -shared -Wl,-soname,libtdjni.so -o libtdjni.so CMakeFiles/tdjni.dir/td_jni.cpp.o /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so /usr/local/lib/libtdclient.a ../td/lib/libtddb.a -ltdcore /opt/jdk1.8.0_101/jre/lib/amd64/server/libjvm.so -Wl,-rpath,/opt/jdk1.8.0_101/jre/lib/amd64/server:`

symbol  _ZN2td16verbosity_binlogE is still not included and not defined.

`nm  ../td/lib/libtddb.a | grep verbosity_binlogE
U _ZN2td16verbosity_binlogE
0000000000000000 D _ZN2td16verbosity_binlogE`

```
nm td/example/java/build/libtdjni.so | grep binlogE
                 U _ZN2td16verbosity_binlogE
0000000000b642f0 t _ZNSt14_Function_base13_Base_managerIZN2td12_GLOBAL__N_111init_binlogERNS1_6BinlogESsRNS1_14BinlogKeyValueIS3_EES7_RNS1_4TdDb6EventsENS1_5DbKeyEEUlRKNS1_11BinlogEventEE_E10_M_managerERSt9_Any_dataRKSH_St18_Manager_operation
0000000000b6cf60 t _ZNSt17_Function_handlerIFvRKN2td11BinlogEventEEZNS0_12_GLOBAL__N_111init_binlogERNS0_6BinlogESsRNS0_14BinlogKeyValueIS6_EESA_RNS0_4TdDb6EventsENS0_5DbKeyEEUlS3_E_E9_M_invokeERKSt9_Any_dataS3_
0000000000b6ccc0 t _ZZN2td12_GLOBAL__N_111init_binlogERNS_6BinlogESsRNS_14BinlogKeyValueIS1_EES5_RNS_4TdDb6EventsENS_5DbKeyEENKUlRKNS_11BinlogEventEE_clESC_
00000000011e33f0 t _ZZN2td18SecretChatsManager24make_secret_chat_contextEiEN7Context6binlogEv
```

**Comment by levlam** [comment]
@UristMcKerman This looks like a bug of the linker, which doesn't include the symbol at offset 0. Could you try to switch to another linker, gold or bfd?

**Comment by UristMcKerman** [comment]
Finally managed to flawlessly assemble and launch java demo on Ubuntu 18 following your guide but still unable to build it on CentOS 7. I think I would abandon any further attempts. Should I close the issue?

P.S. all devs who are like me trying to build it for CentOS 7 listen my words of wisdom: 'JUST DONT'

**Comment by levlam** [comment]
Could you try to workaround the linker bug?

For example by adding at https://github.com/tdlib/td/blob/92c9149e30318ac83baeb194ba9fcacf2bea3588/tddb/td/db/binlog/BinlogEvent.cpp#L15 
```
int32 VERBOSITY_NAME(fake_binlog) = VERBOSITY_NAME(DEBUG) + 8;
```
which will be placed at offset 0, but not used anywhere, so it will not be really needed by the linker.
